net/portmapper/upnp.go

```diff
@@ -0,0 +1,195 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+package portmapper
+
+import (
+	"context"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+	"inet.af/netaddr"
+	"tailscale.com/tempfork/upnp/dcps/internetgateway2"
+)
+
+type upnpMapping struct {
+	gw       netaddr.IP
+	external netaddr.IPPort
+	internal netaddr.IPPort
+	useUntil time.Time
+	client   upnpClient
+}
+
+func (u *upnpMapping) isCurrent() bool                { return u.useUntil.After(time.Now()) }
+func (u *upnpMapping) validUntil() time.Time          { return u.useUntil }
+func (u *upnpMapping) externalIPPort() netaddr.IPPort { return u.external }
+func (u *upnpMapping) release() {
+	u.client.DeletePortMapping(context.Background(), "", u.external.Port(), "udp")
+}
+
+type upnpClient interface {
```

Comment:
Why is this an interface?

Go generally only uses interfaces for something you consume, not something you provide.

Reply:
This is to abstract over the upnp client provided by the fork of the goupnp client, which directly exposes each of the protocols. Should I move the interface into that package?